client/src/interface.rs

```diff
@@ -569,4 +632,231 @@ impl Stronghold {
     fn check_config_flags() {
         unimplemented!()
     }
+
+    #[cfg(feature = "communication")]
+    ///  Start listening on the swarm
+    pub async fn start_listening(&self, addr: Option<Multiaddr>) -> ResultMessage<Multiaddr> {
+        match ask::<_, _, CommunicationResults<SHResults>, _>(
+            &self.system,
+            &self.communication_actor,
+            CommunicationRequest::StartListening(addr),
+        )
+        .await
+        {
+            CommunicationResults::StartListeningResult(Ok(addr)) => ResultMessage::Ok(addr),
+            CommunicationResults::StartListeningResult(Err(_)) => ResultMessage::Error("Listener Error".into()),
+            _ => ResultMessage::Error("Received invalid communication actor response".into()),
+        }
+    }
+
+    #[cfg(feature = "communication")]
+    ///  Get the peer id and listening addresses of the local peer
+    pub async fn get_swarm_info(&self) -> ResultMessage<(PeerId, Vec<Multiaddr>)> {
+        match ask::<_, _, CommunicationResults<SHResults>, _>(
+            &self.system,
+            &self.communication_actor,
+            CommunicationRequest::GetSwarmInfo,
+        )
+        .await
+        {
+            CommunicationResults::SwarmInfo { peer_id, listeners } => ResultMessage::Ok((peer_id, listeners)),
+            _ => ResultMessage::Error("Failed to obtain swarm information".into()),
+        }
+    }
+
+    #[cfg(feature = "communication")]
+    /// Write to the vault of a remote stronghold
+    pub async fn write_remote_vault(
+        &self,
+        peer_id: PeerId,
+        addr: Multiaddr,
+        location: Location,
+        payload: Vec<u8>,
+        hint: RecordHint,
+        _options: Vec<VaultFlags>,
+    ) -> StatusMessage {
+        // connect remote peer
+        match ask::<_, _, CommunicationResults<SHResults>, _>(
+            &self.system,
+            &self.communication_actor,
+            CommunicationRequest::ConnectPeer { peer_id, addr },
+        )
+        .await
+        {
+            CommunicationResults::ConnectPeerResult(Ok(_)) => {}
+            CommunicationResults::ConnectPeerResult(Err(reason)) => {
+                return StatusMessage::Error(format!("Error connecting peer: {:?}", reason))
+            }
+            _ => return StatusMessage::Error("Invalid communication event".into()),
+        };
+
+        let vault_path = &location.vault_path();
+        let vault_path = vault_path.to_vec();
+        // check if vault exists
+        let b = match self
```

Comment:
can you rename these `b` variables to `vault_exists` and `record_exists`?

Reply:
also, should this whole logic be a core stronghold message? like a `EnsureVaultExists` and `EnsureRecordExists` - maybe this is useful for consumers too

Reply:
I copied this logical sequence of the message from the interface for targeting the local client, so this question also concerns those methods. But I can imagine that a user might need these dedicated message types, especially with remote strongholds when they don't want to automatically create a new vault if the vault path doesn't exist, but also in a local system.
@tensor-programming what do you think this?

Reply:
I have to wonder if we should even allow a user to create a remote vault.  Or if there should be some kind of policy for restricting that kind of behavior.   Ideally, we would want each user to have access to a select number of vaults based on which paths they know and haven't really considered how we would restrict users inside of a shared stronghold.  Perhaps, its something to think about later. 

For now I think it should be ok to just enable the ability to add/create a new vault if the vault doesn't exist and we can restrict/change that behavior later using policies.

Reply:
Well, i think this type of policy that you are talking about Tensor could maybe be decided by the Firewall, so the Firewall can be configured for each stronghold individually to allow or reject a `SHRequest::CreateNewVault`?